kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -208,3 +351,225 @@ toRepresentation
+                TermLike.DefinedF defined -> TermLike.getDefined defined
+                _ -> 1 + sum termLikeF
+
+    predSize :: Predicate variable -> Int
+    predSize =
+        Recursive.fold $ \(_ :< predF) ->
+            case predF of
+                Predicate.CeilF ceil_ -> 1 + sum (size <$> ceil_)
+                Predicate.EqualsF equals_ -> 1 + sum (size <$> equals_)
+                Predicate.FloorF floor_ -> 1 + sum (size <$> floor_)
+                Predicate.InF in_ -> 1 + sum (size <$> in_)
+                _ -> 1 + sum predF
+
+    assume
+        :: Predicate variable ->
+        StateT (DoubleMap variable) Changed ()
+    assume predicate =
+        State.modify' (assumeEqualTerms . assumePredicate)
+      where
+        assumePredicate =
+            case predicate of
+                PredicateNot notChild ->
+                    -- Infer that the predicate is \bottom.
+                    Lens.over (field @"predMap") $
+                        HashMap.insert notChild makeFalsePredicate
+                _ ->
+                    -- Infer that the predicate is \top.
+                    Lens.over (field @"predMap") $
+                        HashMap.insert predicate makeTruePredicate
+        assumeEqualTerms =
+            case predicate of
+                PredicateEquals t1 t2 ->
+                    case retractLocalFunction (TermLike.mkEquals_ t1 t2) of
+                        Just (Pair t1' t2') ->
+                            Lens.over (field @"termLikeMap") $
+                                HashMap.insert t1' t2'
+                        _ -> id
+                _ -> id
+
+    applyAssumptions
+        ::  Predicate variable
+        ->  StateT (DoubleMap variable) Changed (Predicate variable)
+    applyAssumptions replaceIn = do
+        assumptions <- State.get
+        lift (applyAssumptionsWorker assumptions replaceIn)
+
+    applyAssumptionsWorker
+        :: DoubleMap variable
+        -> Predicate variable
+        -> Changed (Predicate variable)
+    applyAssumptionsWorker assumptions original
+      | Just result <- HashMap.lookup original (predMap assumptions) = Changed result
+
+      | HashMap.null (termLikeMap assumptions') &&
+        HashMap.null (predMap assumptions') = Unchanged original
+
+      | otherwise = (case replaceIn of
+          Predicate.CeilF ceil_ -> Predicate.CeilF <$> traverse
+            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) ceil_
+          Predicate.FloorF floor_ -> Predicate.FloorF <$> traverse
+            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) floor_
+          Predicate.EqualsF equals_ -> Predicate.EqualsF <$> traverse
+            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) equals_
+          Predicate.InF in_ -> Predicate.InF <$> traverse
+            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) in_
+          _ -> traverse (applyAssumptionsWorker assumptions') replaceIn
+        )
```

Comment:
The [layout rule](https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-210002.7) makes the parentheses unnecessary here:

```suggestion
      | otherwise =
        case replaceIn of
            Predicate.CeilF ceil_ -> Predicate.CeilF <$> traverse
              (applyAssumptionsWorkerTerm (termLikeMap assumptions')) ceil_
            Predicate.FloorF floor_ -> Predicate.FloorF <$> traverse
              (applyAssumptionsWorkerTerm (termLikeMap assumptions')) floor_
            Predicate.EqualsF equals_ -> Predicate.EqualsF <$> traverse
              (applyAssumptionsWorkerTerm (termLikeMap assumptions')) equals_
            Predicate.InF in_ -> Predicate.InF <$> traverse
              (applyAssumptionsWorkerTerm (termLikeMap assumptions')) in_
            _ -> traverse (applyAssumptionsWorker assumptions') replaceIn
```

We should also extract `applyAssumptionsWorkerTerm (termLikeMap assumptions')` and `applyAssumptionsWorker assumptions'` to make this block easier to read.